ompi/communicator/comm.c

```diff
@@ -899,6 +975,26 @@ int ompi_comm_split_type (ompi_communicator_t *comm, int split_type, int key,
         return OMPI_SUCCESS;
     }
 
+    /* MPI_COMM_TYPE_HW_GUIDED: Check if 'value' the same at all ranks */
+    if (tmp[4] != -tmp[5]) {
+        if (0 == ompi_comm_rank(comm)) {
+            opal_output(0, "Error: Mismatched info values for MPI_COMM_TYPE_HW_GUIDED");
+        }
+        return OMPI_ERR_BAD_PARAM;
+    }
+
+    /* TODO: Make this better...
+     *
+     * See Example 7.4 in the MPI 4.0 standard for example usage.
+     *
+     * Stage 0: Recognized, but not implemented.
+     * Stage 1: Do better than that
+     */
+    if (MPI_COMM_TYPE_HW_UNGUIDED == global_split_type) {
```

Comment:
A quick and simple solution to this. Add a field to `ompi_communicator_t` after `c_epoch` (there is room for exactly 4 bytes) and store the type used on the communicator creation (inherit from the parent for duplication, and set to undefined otherwise). Then use this type to go down the `ompi_comm_split_type_hw_guided_support` for the next level in the hierarchy. It might work.

Reply:
That's a good idea. I would like to experiment with that as a second effort (after this PR is merged). I'm going to talk with someone today that might have some cycles in the near term to work on it. Otherwise, I might take a crack at it.